File: perf/cases/cpp/containers.cpp
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <unordered_map>
#include <vector>

int main(int argc, char** argv) {
    const std::int64_t n = argc > 1 ? std::atoll(argv[1]) : 300000LL;
    constexpr std::int64_t mod = 1000000007LL;
    std::vector<std::int64_t> values;
    values.reserve(static_cast<std::size_t>(n));
    for (std::int64_t i = 0; i < n; ++i) {
        values.push_back((i * 31LL) % 4096LL);
    }

    std::int64_t total = 0;
    for (const auto value : values) {
        total = (total + value) % mod;
    }

    std::unordered_map<std::int64_t, std::int64_t> counts;
    counts.reserve(4096);
    for (const auto value : values) {
        ++counts[value];
    }

    for (std::int64_t key = 0; key < 4096; ++key) {
        auto iter = counts.find(key);
        if (iter != counts.end()) {
            total = (total + iter->second * (key + 1)) % mod;
        }
    }
    std::cout << total << '\n';
    return 0;
}

File: perf/cases/python/containers.py
#!/usr/bin/env python3
from __future__ import annotations

import sys

MOD = 1_000_000_007


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 300_000
    values: list[int] = []
    for i in range(n):
        values.append((i * 31) % 4096)

    total = 0
    for value in values:
        total = (total + value) % MOD

    counts: dict[int, int] = {}
    for value in values:
        counts[value] = counts.get(value, 0) + 1

    for key in range(4096):
        total = (total + counts.get(key, 0) * (key + 1)) % MOD
    print(total)


if __name__ == "__main__":
    main()
